daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/ExpressionEvaluatingParsers.scala

```diff
@@ -140,7 +141,16 @@ class AssertExpressionEvaluationParser(
       if (testResult) {
         start.setDiscriminator(discrim)
       } else {
-        // The assertion failed. Prepare a failure message etc. in case backtracking ultimately fails from here.
+        val msg =
+          try {
+            msgExpr.evaluate(start)
+          } catch {
+            case e @ (_: ParseError | _: SchemaDefinitionDiagnosticBase) => {
```

Comment:
Same issue here. Should this be catching SDE?

Also, can't this code, which is common to both Pattern and Expression based Asserts, be factored into a base trait?